TCPStore add watchKey method and new listener thread

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54264 TCPStore add watchKey method and new listener thread**

**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

Differential Revision: [D27709912](https://our.internmc.facebook.com/intern/diff/D27709912)

H-Huang committed:
Add watch_key query and listen thread outline

[ghstack-poisoned]

H-Huang committed:
Update on "Add watch_key query and listen thread outline"




[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -428,17 +455,26 @@ TCPStore::TCPStore(
+      // socket to handle requests from server
+      listenSocket_ = tcputil::connect(
+          tcpStoreAddr_, tcpStorePort_, /* wait= */ true, timeout_);
+      tcpStoreListener_ = std::make_unique<ListenThread>(listenSocket_);
```
do we plan to use this listener to handle things other than watch callback? If not, does it make sense to call it `watchListner_` to be more specific?

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -516,6 +552,15 @@ bool TCPStore::deleteKey(const std::string& key) {
+void TCPStore::watchKey(const std::string& key, CallbackFunction callback) {
+  std::string regKey = regularPrefix_ + key;
+
+  tcpStoreListener_->addCallback(regKey, callback);
```
Would I be correct if I assume the thread-safety is based on the assumption that callback function itself does not call any TCPStore APIs? For example, if I register a callback that will register another callback when fired, will it still be thread-safe? Probably not because they are called in two different threads (main thread and the `tcpStoreListener_` thread)? It's totally fine to keep it so for simplicity for now, as we fully control the callback implementation in the foreseeable future. But let's highlight that risk in the `addCallback`/`CallbackFunction` API comments.

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -61,8 +63,24 @@ class TCPStoreDaemon {
 #endif
 };
 
+class ListenThread {
```
Shall we add comments to this class and it's member functions and variables?

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -61,8 +63,24 @@ class TCPStoreDaemon {
 
+class ListenThread {
+  public:
+    typedef void (*CallbackFunction)(const std::string&, const std::string&);
```
I don't have strong preferences on this. Just wanna point out that existing code uses either dedicated struct/class or `std::function`:

1. dedicated class:

https://github.com/pytorch/pytorch/blob/19792b45dbf30b4555c4a87512e624cdd4aa6e4c/torch/csrc/distributed/rpc/request_callback.h

https://github.com/pytorch/pytorch/blob/19792b45dbf30b4555c4a87512e624cdd4aa6e4c/torch/lib/c10d/comm.hpp#L79

2. std::function (the caller can capture whatever they need in the function):

https://github.com/pytorch/pytorch/blob/19792b45dbf30b4555c4a87512e624cdd4aa6e4c/aten/src/ATen/core/ivalue_inl.h#L449

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/CUDAFuture.h#L74

H-Huang replied:
Thanks for the note, changed to use `std::function`

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -61,8 +63,24 @@ class TCPStoreDaemon {
+
 class TCPStore : public Store {
  public:
+  typedef void (*CallbackFunction)(const std::string&, const std::string&);
```
this is duplicated with the typedef above? Can this be a global typedef in c10d namespace?

mrshenli replied:
If this is specific to the watch event, does it make sense to name it `WatchCallback`?

mrshenli reviewed (commented):
In general LGTM! I will hold the stamp until I finish reading the full stack and tests.

pritamdamania87 commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -428,17 +455,26 @@ TCPStore::TCPStore(
       }
+
+      // socket to handle requests from server
+      listenSocket_ = tcputil::connect(
```
Can't we just reuse `storeSocket_` instead of having another socket for the same addr and port?

H-Huang replied:
We could do this, but that would involve rewriting all the current methods and some util methods to be asynchronous, since there is currently no way do something like execute a callback while in the middle of `wait()`. To get around this we created a dedicated listen thread which reads from its own socket to handles such asynchronous methods.

pritamdamania87 commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -86,6 +104,8 @@ class TCPStore : public Store {
 
   bool deleteKey(const std::string& key) override;
 
+  void watchKey(const std::string& key, CallbackFunction callback);
```
We probably need some docs for this function somewhere in terms of when CallbackFunction is called and with what args. Also, don't we need to define this in Store.hpp and override it here?

H-Huang replied:
Thanks, I will make these changes

pritamdamania87 reviewed (commented):
It would be good to add some summary to the PR describing the overall changes.

H-Huang committed:
Update on "Add watch_key query and listen thread outline"




[ghstack-poisoned]

H-Huang committed:
Update on "Add watch_key query and listen thread outline"




[ghstack-poisoned]

H-Huang committed:
Update on "Add watch_key query and listen thread outline"




[ghstack-poisoned]

H-Huang committed:
Update on "Add watch_key query and listen thread outline"


**Changes**
This PR adds the methods and classes for the listener thread and watch key query. It also adds the method to prefix store. Implementation of these methods are in the next PR.

**Background**
This stack of PRs adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and the `callback(old_key, new_key)` function to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and Rref context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -57,10 +58,154 @@ TCPStoreDaemon::~TCPStoreDaemon() {
+}
+
+void ListenThread::addCallback(std::string key, std::function<void(std::string, std::string)> cb) {
+  keyToCallbacks_[key] = cb;
```
Don't we need a lock for `keyToCallbacks_`? `addCallback` will be called by the user thread and at the same time `ListenThread::callbackHandler` can be called from the background thread which again accesses `keyToCallbacks_`

pritamdamania87 commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -164,11 +326,30 @@ void TCPStoreDaemon::wakeupWaitingClients(const std::string& key) {
-  tcpStore_[key] = tcputil::recvVector<uint8_t>(socket);
+  std::vector<uint8_t> newData = tcputil::recvVector<uint8_t>(socket);
+  std::vector<uint8_t> oldData;
+  if (tcpStore_.find(key) != tcpStore_.end()) {
+    oldData = tcpStore_.at(key);
```
Avoid extra map lookup as follows:

```
auto it = tcpStore_.find(key);
if (it != tcpStore_.end()) {
 oldData = *it;
 tcpStore_.insert_or_assign(it, key, newData);
} else {
 tcpStore_[key] = newData;
}
```

pritamdamania87 commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -193,17 +377,22 @@ void TCPStoreDaemon::addHandler(int socket) {
 
+  std::vector<uint8_t> oldData;
   if (tcpStore_.find(key) != tcpStore_.end()) {
+    oldData = tcpStore_[key];
```
Same as my suggestion in `setHandler`.

pritamdamania87 commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -220,6 +409,8 @@ void TCPStoreDaemon::deleteHandler(int socket) {
   auto numDeleted = tcpStore_.erase(key);
   tcputil::sendValue<int64_t>(socket, numDeleted);
+  // Remove all clients watching the key
+  watchedSockets_.erase(key);
```
Is this behavior correct? What if someone sets the same key after the delete? Shouldn't the user get notified of that?

pritamdamania87 reviewed (changes requested):
We need some comprehensive unit tests with this PR to ensure watchKey is working as expected for different combinations of operations.

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

Differential Revision: [D27709912](https://our.internmc.facebook.com/intern/diff/D27709912)

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

Differential Revision: [D27709912](https://our.internmc.facebook.com/intern/diff/D27709912)

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

Differential Revision: [D27709912](https://our.internmc.facebook.com/intern/diff/D27709912)

[ghstack-poisoned]

mrshenli reviewed (commented):
1st batch, mostly minor ones. Will come back to this later today.

H-Huang commented on torch/lib/c10d/PrefixStore.hpp:
```diff
@@ -40,6 +40,9 @@ class PrefixStore : public Store {
 
   void setTimeout(const std::chrono::milliseconds& timeout) override;
 
+  void watchKey(const std::string& key,
+    std::function<void(c10::optional<std::string>, c10::optional<std::string>)> callback) override;
```
will do

H-Huang commented on torch/lib/c10d/Store.hpp:
```diff
@@ -55,6 +55,13 @@ class Store : public torch::CustomClassHolder {
 
+  virtual void watchKey(const std::string& key,
+    std::function<void(c10::optional<std::string>, c10::optional<std::string>)> callback) {
+        std::ignore = key;
+        std::ignore = callback;
```
Thanks! Good to know

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

Differential Revision: [D27709912](https://our.internmc.facebook.com/intern/diff/D27709912)

[ghstack-poisoned]

H-Huang commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -1,15 +1,15 @@
 
-#include <algorithm>
 #include <fcntl.h>
+#include <algorithm>
```
I think this was from run clang-format previously

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -14,17 +14,42 @@
-  explicit TCPStoreDaemon(int storeListenSocket);
-  ~TCPStoreDaemon();
+  explicit BackgroundThread(int storeListenSocket);
+  virtual ~BackgroundThread() = 0;
```
I am a little confused. This is a pure virtual function, but we still provided implementation for it? C++ reference mentioned that:

> C++ does not prohibit Base from providing a definition for the pure virtual function. However, the derived class still has to implement the pure virtual function. Definition of the pure virtual function has to be placed outside the declaration (in header or source file - doesn't matter).

https://en.cppreference.com/book/intro/abstract_classes

H-Huang replied:
Yep, the references seems good, a pure virtual function only **requires** derived classes to override the method, which is orthogonal to whether it is implemented by the base class. The pure virtual destructor prevents the class from being instantiated and the derived classes implicitly override it with their default destructors. The destructors still gets called in the normal order, ~derived() -> ~base()

I just wanted to make `BackgroundThread` an abstract class since C++ doesn't have the `abstract` keyword, if there are any other ways we usually do it let me know!

mrshenli replied:
Trying to understand this :)

I didn't see a destructor in `TCPStoreMasterDaemon`, and there is no problem for creating `TCPStoreMasterDaemon` objects. Is it because C++ considers the implicit destructor of `TCPStoreMasterDaemon` overrides `BackgroundThread`'s pure virtual destructor?

H-Huang replied:
Yes, exactly!

mrshenli commented on torch/lib/c10d/Store.hpp:
```diff
@@ -55,6 +55,14 @@ class Store : public torch::CustomClassHolder {
 
   virtual void setTimeout(const std::chrono::milliseconds& timeout);
 
+  virtual void watchKey(
```
add some comments here to explain what derived classes should do if they plans to implement this API?

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -14,17 +14,42 @@
+  std::vector<int> controlPipeFd_{-1, -1};
+#endif
+ private:
   void join();
+  void stop();
+  void initStopSignal();
+  void closeStopSignal();
```
this is prior to this PR. Shall we add comments to explain each method and major field in this class?

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -57,10 +56,165 @@ TCPStoreDaemon::~TCPStoreDaemon() {
+    std::string key,
+    std::function<void(c10::optional<std::string>, c10::optional<std::string>)>
+        cb) {
+  keyToCallbacksLock.lock();
```
other part of the code base usually use a `lock_guard`. See:

https://github.com/pytorch/pytorch/blob/2bb58a06ef845ec6adf632415857cfc0c04cad9b/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L989

H-Huang replied:
Thanks! This is much better

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -36,29 +61,49 @@ class TCPStoreDaemon {
-#endif
+// Listener thread runs on all processes
+// Right now only handles callbacks registered from watchKey()
+class ListenThread : public BackgroundThread {
```
Is this specifically added for watch callbacks? If so, does it make sense to use the following class names?

BackgroundThread -> Daemon
ListenThread -> WatchCallbackDaemon

mrshenli replied:
Sorry, I missed this:

> Right now only handles callbacks registered from watchKey()

If we plan to add more features into this in the future, does it make sense to name them 

TCPStoreDaemon -> TCPStoreMasterDaemon
ListenThread -> TCPStoreWorkerDaemon

H-Huang replied:
Good idea, I will rename it to these.

mrshenli commented on torch/lib/c10d/test/TCPStoreTest.cpp:
```diff
@@ -9,20 +9,30 @@
 #include <c10d/TCPStore.hpp>
 
+std::condition_variable cv;
+std::mutex cv_m;
```
other vars uses camel instead of snake naming?

mrshenli commented on torch/lib/c10d/TCPStore.hpp:
```diff
@@ -114,7 +166,9 @@ class TCPStore : public Store {
   int storeSocket_ = -1;
+  int listenSocket_ = -1;
   int masterListenSocket_ = -1;
+  std::thread listenThread_;
```
is this var used anywhere?

H-Huang replied:
This was old, thanks!

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -57,10 +56,165 @@ TCPStoreDaemon::~TCPStoreDaemon() {
   daemonThread_.join();
 }
 
+#ifdef _WIN32
```
(can be done in follow up PRs) to get rid of frequent `#ifdef`, we can have two implementations for Unix and Windows respectively. `UnixSockUtils.hpp` and `WinSockUtils.hpp` can serve as an example.

H-Huang commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -260,30 +466,20 @@ void TCPStoreDaemon::waitHandler(int socket) {
+  std::string key = tcputil::recvString(socket);
+
+  // record the socket to respond to when the key is updated
+  watchedSockets_[key].push_back(socket);
```
Yep that is correct, only one handler is executed at a time. I will update the hpp file with comments.

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

Differential Revision: [D27709912](https://our.internmc.facebook.com/intern/diff/D27709912)

[ghstack-poisoned]

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

Differential Revision: [D27709912](https://our.internmc.facebook.com/intern/diff/D27709912)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/Store.hpp:
```diff
@@ -55,6 +55,30 @@ class Store : public torch::CustomClassHolder {
 
   virtual void setTimeout(const std::chrono::milliseconds& timeout);
 
+  /*
```
nit: comments in other parts of the code base usually use // instaed of /**/, even for multi-line comments.

mrshenli reviewed (commented):
Will come back to this later today.

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

Differential Revision: [D27709912](https://our.internmc.facebook.com/intern/diff/D27709912)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -57,11 +56,166 @@ TCPStoreDaemon::~TCPStoreDaemon() {
+  std::vector<uint8_t> newValueVec = tcputil::recvVector<uint8_t>(socket);
+  c10::optional<std::string> currentValue;
+  if (watchResponse == WatchResponseType::KEY_CREATED) {
+    currentValue = c10::nullopt;
```
do we need to assert that `currentValueVec` is empty?

mrshenli commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -57,11 +56,166 @@ TCPStoreDaemon::~TCPStoreDaemon() {
+  }
+  c10::optional<std::string> newValue;
+  if (watchResponse == WatchResponseType::KEY_DELETED) {
+    newValue = c10::nullopt;
```
ditto for `newValueVec`

mrshenli reviewed (approved):
LGTM! Thanks for adding this feature!

H-Huang committed:
Update on "TCPStore add watchKey method and new listener thread"


**Changes**

- Creates new listener thread on each client to run the callback
- Create new class which listener thread and master thread derive from, this class is used to handle shut down and clean up of the thread in windows and linux
- Add watchKey method and update any functions that changes the key value.

**Background**
This PR adds functionality to TCPStore to allow users to watch a key and execute a callback on key change.

It introduces this a new watchKey() API:
`TCPStore::watchKey(const std::string& key, std::function<void(std::string, std::string)> callback)` which has parameters `key` and `callback(old_key, new_key)` to run on key change. Since current methods are blocking, for example in`TCPStore::get()` a worker will send a "get key" request to the master -> wait for a response back -> then exit the function and return the value to user, we need a non-blocking, asynchronous way to execute the callback whenever a key changes. This is done by creating a new listener thread on each client which the master can communicate with.


Right now, the API is C++ only and only for TCPStore, the internal use case is for elastic RPC. We will have an internal key such as `_NumNodes` and all nodes in the elastic RPC group will watch this key. When a node leaves, this key will be updated and each node will execute a callback to clean up Autograd context and RRef context.

Differential Revision: [D27709912](https://our.internmc.facebook.com/intern/diff/D27709912)

[ghstack-poisoned]

mrshenli commented:
`testWatchKeyCallback` is very flaky, shall we revert?

H-Huang commented:
Reverting due to flaky tests on 3 workflows:
https://app.circleci.com/pipelines/github/pytorch/pytorch/302262/workflows/d16c584c-0ef4-4978-896e-9677c8657274/jobs/12439535
https://app.circleci.com/pipelines/github/pytorch/pytorch/302262/workflows/d16c584c-0ef4-4978-896e-9677c8657274/jobs/12439533
https://app.circleci.com/pipelines/github/pytorch/pytorch/302134/workflows/a702ab71-7b1e-4b0d-8e97-10094a89fc4a/jobs/12433543
